torchtitan/utils.py

```diff
@@ -52,26 +53,75 @@ def _warn_overwrite_env(env, val):
+    seed = job_config.training.seed
+
+    if deterministic:
+        logger.info("Deterministic training enabled (expect perf degradation).")
+        torch.use_deterministic_algorithms(True)
+        # env var for deterministic CuBLAS
+        # https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
+    torch.backends.cudnn.deterministic = deterministic
+    torch.backends.cudnn.benchmark = not deterministic
+
+    # to ensure we can control which ranks have same or different seeds, all ranks agree on a starting seed.
+    # if user provides one, we use this. Otherwise rank 0 rolls the dice and everyone else uses that.
     if seed is not None:
         # CPU and GPU determinism
         torch.manual_seed(seed)
-        # set deterministic cudnn algorithms
-        torch.backends.cudnn.deterministic = True
-        torch.backends.cudnn.benchmark = False
         # set Python seed
         os.environ["PYTHONHASHSEED"] = str(seed)
-        torch.use_deterministic_algorithms(True)
-        # env var for deterministic CuBLAS
-        # https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
-        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
+
     else:
-        # ensure we turn off deterministic cudnn algorithms
-        torch.backends.cudnn.deterministic = False
-        torch.backends.cudnn.benchmark = True
+        # Extract the seed for torch's main generator on rank 0 and standardizes on using that to build
+        # seeds for unique SPMD groups
+        seed_tensor = torch.get_rng_state()[:8].to(device)
+        torch.distributed.broadcast(seed_tensor, src=0)
+        seed = seed_tensor.view(torch.uint64).item()
+
+    # In special cases the trainer can be launched with a single rank, in which case we're done now
+    if c10d.get_world_size() == 1:
+        return
+
+    # For PP + SPMD cases, we want to separate the world into the SPMD mesh and the PP mesh,
+    # and choose a unique seed for each rank on the PP mesh.
+    if "pp" in world_mesh.mesh_dim_names:
+        pp_mesh = world_mesh["pp"]
+        seed += pp_mesh.get_local_rank()
+        seed %= 2**64 - 1
+
+        logger.debug(
+            f"PP rank {pp_mesh.get_local_rank()}, Global rank {c10d.get_rank()} using seed: {seed}"
+        )
+        spmd_mesh_dims = list(
+            filter(lambda name: name != "pp", world_mesh.mesh_dim_names)
+        )
+        spmd_mesh = world_mesh[spmd_mesh_dims] if len(spmd_mesh_dims) else None
+    else:
+        spmd_mesh = world_mesh
+        logger.debug(f"Global Rank {c10d.get_rank()} using seed: {seed}")
+
+    # As long as we are not in the 1-D (PP-only) case, we will have a seed to use for all ranks of the SPMD mesh.
+    # IF PP is also used, this seed is unique per PP rank.
+    if spmd_mesh:
+        torch.distributed.tensor._random.manual_seed(seed, spmd_mesh)
```

Comment:
oops, i think i need to call torch.manual_seed()  again here, at least in the non-spmd case- to ensure the updated pp seed is used.

Reply:
how about when `spmd_mesh is None`, do different PP ranks get different `torch.manual_seed(seed)`?

Reply:
or call `torch.manual_seed` no matter if `spmd` exists? It doesn't hurt in SPMD case since we use DTensor.

Reply:
yea, i have made it so that manual_seed (and python seed, fwiw) are called no matter what.